subscription service with test of exception

mrE100 opened the pull request.

mrE100 committed:
subscription service with test of exception

mrE100 committed:
subscription service WITH NO TEST

mrE100 committed:
subscription service WITH NO TEST

mrE100 committed:
SOME tests added

three-moon commented on src/main/java/school/faang/user_service/mapper/UserMapper.java:
```diff
@@ -0,0 +1,18 @@
+
+    User toEntity(UserDto userDto);
+
+    List<UserDto> toDto(List<User> list);
```
Мне кажется лучше назвать метод toListDto

Vahe650 replied:
пригодится еще метод такой
List<User> toEntity(List<UserDto> list);

avfedorenko commented on src/test/java/school/faang/user_service/service/SubscriptionServiceTest.java:
```diff
@@ -0,0 +1,89 @@
+@ExtendWith(MockitoExtension.class)
+public class SubscriptionServiceTest {
+    private Long userId1 = 1000L;
+    private Long userId2 = 2000L;
```
Это наверное лучше вынести в отдельный метод под аннотацию BeforeEach

avfedorenko commented on src/test/java/school/faang/user_service/service/SubscriptionServiceTest.java:
```diff
@@ -0,0 +1,89 @@
+    private SubscriptionRepository subscriptionRepository;
+
+//    @Spy
+//    private UserMapper userMapper = Mappers.getMapper(UserMapper.class);
```
Можно заменить на private UserMapperImpl userMapper; и раскомментировать

avfedorenko replied:
Если не получится импортировать сразу, то попробуй запустить проект один раз

avfedorenko commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,67 @@
+        return subscriptionRepository.findFolloweesAmountByFollowerId(followerId);
+    }
+
+    private List<UserDto> filterUser(Stream<User> users, UserFilterDto filters) {
```
В задании название метода было filterUsers, т.к несколько пользователей обрабатываем тут

avfedorenko commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    public List<UserDto> getFollowing(long followerId, UserFilterDto filter) {
+        Stream<User> users = subscriptionRepository.findByFolloweeId(followerId);
```
Мне кажется здесь должен быть метод findByFollowerId потому что мы ищем тех на кого юзер подписаны, то есть у нас конкретный follower(кто подписался) для разных followee(на кого подписались). Там в условии так и написано как у тебя сделано, но я не согласен с условием.

avfedorenko commented on src/test/java/school/faang/user_service/service/SubscriptionServiceTest.java:
```diff
@@ -0,0 +1,89 @@
+//        user1.setId(userId1);
+//        user1.setFollowers();
+//        when(subscriptionRepository.findByFollowerId(userId1)).thenReturn(user1.getFollowers().stream());
+//
```
@Test
    public void testGetFollowers() {
        User user1 = new User();
        user1.setId(userId1);
        user1.setFollowers(List.of(User.builder().id(userId2).build()));
        when(subscriptionRepository.findByFolloweeId(userId1)).thenReturn(user1.getFollowers().stream());
        List<UserDto> result = subscriptionService.getFollowers(userId1, new UserFilterDto());
        assertEquals(result.get(0).getId(), userId2);
    }
    Этот тест может выглядеть так например. Про путаницу с follower/followee я выше расписал

avfedorenko commented on src/test/java/school/faang/user_service/service/SubscriptionServiceTest.java:
```diff
@@ -0,0 +1,89 @@
+
+    @InjectMocks
+    private SubscriptionService subscriptionService;
+
```
private final List<UserFilter> userFilters = new ArrayList<>();

    private static long userId1;
    private static long userId2;

    @BeforeEach
    public void setUp() {
        userId1 = 1000L;
        userId2 = 2000L;
        userFilters.add(new NamePatternFilter());
        subscriptionService = new SubscriptionService(subscriptionRepository, userMapper, userFilters);
    }
    Сюда (после моков перед тестами) я бы вставил это, так как без инициализации фильтров вручную они сами работать не будут, а чтобы автоматически было - не нашел решение. Теперь переменные с id пользователей объявляются перед каждым тестом заново. Выше написал откуда я бы их убрал и перенес сюда

mrE100 committed:
added tests

Vahe650 commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.List;
+
+
+@Component
```
@RestController

Vahe650 commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.List;
+import java.util.stream.Stream;
+
+@Component
```
@Service

mrE100 replied:
Это добавить аннотацию или заменить?

Vahe650 replied:
заменить

Vahe650 commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,67 @@
+public class SubscriptionService {
+    private final SubscriptionRepository subscriptionRepository;
+    private final UserMapper userMapper;
+    private final List<UserFilter> userFilters;
```
тут лучше инжектить все бины с анотацией @Qualifier чтобы визуально можно было увидеть какие действия выполняются

например для AboutPatternFilter инжект будет выглятеть так

@Qualifier("aboutPatternFilter") 
private final UserFilter aboutPatternFilter

mrE100 replied:
То есть просто добавить к каждому фильтру своё поле с аннотацией? А конструктор тогда как сделать, чтобы если не указано поле, null ставилось? Есть ли автомат для этого, или вручную через ифы перебор делать?

Vahe650 replied:
Какой конструктор?

mrE100 replied:
в тестах создаем несколько фильтров и добавляем их в лист. А если фильтра какого-то нет, можно ему по-умолчанию поставить null?

mrE100 replied:
у нас же инжектятся все бины с фильтрами автоматом, чтобы была расширяемость. А если @qualifier навесить, то придётся каждый вручную прописывать, если что

Vahe650 replied:
возможно придется пару строчек вручную прописать , но визуально твой код будет более читабельным и ревюер сразу поймет какую работу выполняет твой код

Vahe650 replied:
я не говорю что ты сделал ошыбку, просто на этом этапе давай сделаем через Qualifier.
заодно и ознакомишься с анотацией

Vahe650 commented on src/main/java/school/faang/user_service/service/filters/AboutPatternFilter.java:
```diff
@@ -0,0 +1,18 @@
+
+public class AboutPatternFilter implements UserFilter {
+    @Override
+    public boolean isApplicable(UserFilterDto filters) {
```
public boolean isApplicable(UserFilterDto userFilterDto) {

Vahe650 commented on src/main/java/school/faang/user_service/service/filters/CityPatternFilter.java:
```diff
@@ -0,0 +1,18 @@
+
+public class CityPatternFilter implements UserFilter {
+    @Override
+    public boolean isApplicable(UserFilterDto filters) {
```
public boolean isApplicable(UserFilterDto userFilterDto) {

Vahe650 commented on src/main/java/school/faang/user_service/service/filters/CityPatternFilter.java:
```diff
@@ -0,0 +1,18 @@
+    }
+
+    @Override
+    public void apply(Stream<User> users, UserFilterDto filters) {
```
public void apply(Stream<User> users, UserFilterDto userFilterDto) {

Vahe650 commented on src/main/java/school/faang/user_service/service/filters/ContactPatternFilter.java:
```diff
@@ -0,0 +1,18 @@
+    }
+
+    @Override
+    public void apply(Stream<User> users, UserFilterDto filters) {
```
userFilterDto

Vahe650 commented on src/main/java/school/faang/user_service/service/filters/ContactPatternFilter.java:
```diff
@@ -0,0 +1,18 @@
+
+public class ContactPatternFilter implements UserFilter {
+    @Override
+    public boolean isApplicable(UserFilterDto filters) {
```
userFilterDto

Vahe650 commented on src/main/java/school/faang/user_service/service/filters/CountryPatternFilter.java:
```diff
@@ -0,0 +1,18 @@
+
+public class CountryPatternFilter implements UserFilter {
+    @Override
+    public boolean isApplicable(UserFilterDto filters) {
```
userFilterDto

Vahe650 commented on src/main/java/school/faang/user_service/service/filters/CountryPatternFilter.java:
```diff
@@ -0,0 +1,18 @@
+    }
+
+    @Override
+    public void apply(Stream<User> users, UserFilterDto filters) {
```
userFilterDto

Vahe650 commented on src/test/java/school/faang/user_service/service/SubscriptionServiceTest.java:
```diff
@@ -0,0 +1,116 @@
+    public void initialize() {
+        userId1 = 1000L;
+        userId2 = 2000L;
+        userFilters.add(new NamePatternFilter());
```
добавь еще несколько фильтров плз

mrE100 committed:
some annotatiions added

mrE100 committed:
annotation component changed to service and controller

mrE100 committed:
userdto changed to SubscriptionUserDto to avoid conflicts

Commit:
fix some bugs and adding ControllerTest

# Conflicts:
#	src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java
#	src/main/java/school/faang/user_service/mapper/StringToLongMapper.java

mrE100 committed:
Merge branch 'dragon-master-bc3' into dragon-master-bc3-feature-BJS2-3305

mrE100 committed:
conflicts resolving

mrE100 committed:
some Dto files

mrE100 committed:
some Dto files

mrE100 committed:
some Dto files

mrE100 committed:
some Dto files

mrE100 committed:
some Dto files

mrE100 committed:
some

mrE100 committed:
Merge branch 'dragon-master-bc3' into dragon-master-bc3-feature-BJS2-3305

mrE100 committed:
Merge branch 'dragon-master-bc3' into dragon-master-bc3-feature-BJS2-3305
